File: __fixtures__/java/com/acme/web/DonationController.java
package com.acme.web;

import com.acme.domain.Donation;
import com.acme.service.DonationService;
import java.util.List;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/donations")
public class DonationController {

    @Autowired
    private DonationService donationService;

    @GetMapping
    public List<Donation> list() {
        return donationService.listAll();
    }

    @PostMapping
    public Donation create(String donorName, int amount) {
        Donation saved = donationService.record(donorName, amount);
        return saved;
    }

    @GetMapping("/total")
    public int total() {
        return donationService.totalRaised();
    }
}

File: __fixtures__/ts/src/web/donationController.ts
// Imported through the barrel file rather than the defining module.
import { Donation, DonationRepository } from '../domain';
import { DonationService } from '../service/donationService';
import { InMemoryDonationRepository } from '../repo/donationRepository';

export class DonationController {
  private service: DonationService;

  constructor() {
    const repository: DonationRepository = new InMemoryDonationRepository();
    this.service = new DonationService(repository);
  }

  list(): Donation[] {
    return this.service.listAll();
  }

  create(donorName: string, amount: number): string {
    const donation = this.service.record(donorName, amount);
    return donation.describe();
  }

  total(): number {
    return this.service.totalRaised();
  }

  // Donation reaches this file through `export *` in the barrel.
  fromRaw(amount: number, donorName: string): string {
    const donation = new Donation(amount, donorName);
    return donation.describe();
  }

  // DonationRepository reaches this file through `export { X } from` in the barrel.
  seed(repository: DonationRepository): void {
    repository.save(new Donation(0, 'seed'));
  }
}
